src/mpid/ch4/generic/am/mpidig_am_init.c

```diff
@@ -8,21 +8,78 @@
 #include "mpidch4r.h"
 #include "mpidu_genq.h"
 
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_CH4_AM_PACK_BUFFER_SIZE
+      category    : CH4
+      type        : int
+      default     : 69632
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_LOCAL
+      description : >-
+        Specifies the number of buffers for packing/unpacking active messages in
+        each block of the pool. The size here should be greater or equal to the
+        max of the EAGER THRESHOLD and MPIR_CVAR_CH4_SHM_POSIX_IQUEUE_CELL_SIZE.
+        The default set to 17 * 4KB (=MPIR_CVAR_CH4_SHM_POSIX_IQUEUE_CELL_SIZE).
+
+    - name        : MPIR_CVAR_CH4_NUM_AM_PACK_BUFFERS_PER_CHUNK
+      category    : CH4
+      type        : int
+      default     : 16
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_LOCAL
+      description : >-
+        Specifies the number of buffers for packing/unpacking active messages in
+        each block of the pool.
+
+    - name        : MPIR_CVAR_CH4_MAX_AM_UNEXPECTED_PACK_BUFFERS_SIZE_BYTE
+      category    : CH4
+      type        : int
+      default     : 8388608
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_LOCAL
+      description : >-
+        Specifies the max number of buffers for packing/unpacking active messages
+        in the pool.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
 static int dynamic_am_handler_id = MPIDIG_HANDLER_STATIC_MAX;
 
 static void *host_alloc(uintptr_t size);
+static void *host_alloc_buffer_registered(uintptr_t size);
 static void host_free(void *ptr);
+static void host_free_buffer_registered(void *ptr);
 
 static void *host_alloc(uintptr_t size)
 {
     return MPL_malloc(size, MPL_MEM_BUFFER);
 }
 
+static void *host_alloc_buffer_registered(uintptr_t size)
+{
+    void *ptr = MPL_malloc(size, MPL_MEM_BUFFER);
```

Comment:
```suggestion
    void *ptr = MPL_malloc(size, MPL_MEM_BUFFER);
    MPIR_Assert(ptr);
```

Reply:
I think this wouldn't be necessary, as the GenQ already check for NULL pointer and throws an error.

Reply:
But you call buffer registration immediately after that, which won't work on NULL pointers.

Reply:
fixed